File: app/src/main/java/fr/frogdevelopment/pregnancycalendar/ui/chrono/SwipeToDelete.java
package fr.frogdevelopment.pregnancycalendar.ui.chrono;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

import fr.frogdevelopment.pregnancycalendar.R;

import static androidx.recyclerview.widget.ItemTouchHelper.ACTION_STATE_IDLE;
import static androidx.recyclerview.widget.ItemTouchHelper.LEFT;
import static androidx.recyclerview.widget.ItemTouchHelper.RIGHT;

class SwipeToDelete extends ItemTouchHelper.SimpleCallback {

    @FunctionalInterface
    public interface DeleteListener {
        void onDelete(@NonNull RecyclerView.ViewHolder viewHolder);
    }

    private final DeleteListener mDeleteListener;

    // we want to cache these and not allocate anything repeatedly in the onChildDraw method
    private final Drawable mBackground;
    private final Drawable mTrashIcon;
    private final int mIconMargin;

    SwipeToDelete(Context context, DeleteListener deleteListener) {
        super(ACTION_STATE_IDLE, LEFT | RIGHT);
        this.mDeleteListener = deleteListener;

        mBackground = new ColorDrawable(Color.RED);
        mTrashIcon = context.getResources().getDrawable(R.drawable.ic_baseline_delete_sweep_24, null);
        mIconMargin = (int) context.getResources().getDimension(R.dimen.ic_clear_margin);
    }

    @Override
    public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
        // not important, we don't want drag & drop
        return false;
    }

    @Override
    public void onChildDraw(@NonNull Canvas c, @NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, float dX, float dY, int actionState, boolean isCurrentlyActive) {
        super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);

        if (actionState != ItemTouchHelper.ACTION_STATE_SWIPE) {
            return;
        }

        int backgroundCornerOffset = 20;
        int cornerRadius = 30;
        View itemView = viewHolder.itemView;

        int iconMargin = (itemView.getHeight() - mTrashIcon.getIntrinsicHeight()) / 2;
        int iconTop = itemView.getTop() + iconMargin;
        int iconBottom = iconTop + mTrashIcon.getIntrinsicHeight();

        if (dX > 0) { // Swiping to the right
            // Draw background
            mBackground.setBounds(
                    itemView.getLeft(),
                    itemView.getTop(),
                    itemView.getLeft() + (int) dX + backgroundCornerOffset,
                    itemView.getBottom()
            );
            // Draw the delete icon
            mTrashIcon.setBounds(
                    itemView.getLeft() + iconMargin,
                    iconTop,
                    itemView.getLeft() + iconMargin + mTrashIcon.getIntrinsicWidth(),
                    iconBottom
            );
        } else if (dX < 0) { // Swiping to the left
            // Draw background
            mBackground.setBounds(
                    itemView.getRight() + (int) dX - backgroundCornerOffset,
                    itemView.getTop(),
                    itemView.getRight(),
                    itemView.getBottom()
            );
            // Draw the delete icon
            mTrashIcon.setBounds(
                    itemView.getRight() - iconMargin - mTrashIcon.getIntrinsicWidth(),
                    iconTop,
                    itemView.getRight() - iconMargin,
                    iconBottom
            );
        } else { // view is unSwiped
            cornerRadius = 0;
            mBackground.setBounds(0, 0, 0, 0);
            mTrashIcon.setBounds(0, 0, 0, 0);
        }

        mBackground.draw(c);
        mTrashIcon.draw(c);

        if (viewHolder.itemView instanceof CardView) {
            CardView cardView = (CardView) viewHolder.itemView;
            cardView.setRadius(cornerRadius);
        }
    }

    @Override
    public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
        mDeleteListener.onDelete(viewHolder);
    }
}

File: app/src/main/java/fr/frogdevelopment/pregnancycalendar/ui/contraction/SwipeToDelete.kt
package fr.frogdevelopment.pregnancycalendar.ui.contraction

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.ItemTouchHelper.*
import androidx.recyclerview.widget.RecyclerView
import fr.frogdevelopment.pregnancycalendar.R

internal class SwipeToDelete(context: Context, private val mDeleteListener: DeleteListener) : SimpleCallback(ACTION_STATE_IDLE, LEFT or RIGHT) {

    @FunctionalInterface
    interface DeleteListener {
        fun onDelete(viewHolder: RecyclerView.ViewHolder)
    }

    private val mBackground: Drawable
    private val mTrashIcon: Drawable

    override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean { // not important, we don't want drag & drop
        return false
    }

    override fun onChildDraw(c: Canvas, recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, dX: Float, dY: Float, actionState: Int, isCurrentlyActive: Boolean) {
        super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)

        if (actionState != ACTION_STATE_SWIPE) {
            return
        }

        val backgroundCornerOffset = 20
        var cornerRadius = 30
        val itemView = viewHolder.itemView
        val iconMargin = (itemView.height - mTrashIcon.intrinsicHeight) / 2
        val iconTop = itemView.top + iconMargin
        val iconBottom = iconTop + mTrashIcon.intrinsicHeight

        when {
            // Swiping to the right
            dX > 0 -> {
                // Draw background
                mBackground.setBounds(
                        itemView.left,
                        itemView.top,
                        itemView.left + dX.toInt() + backgroundCornerOffset,
                        itemView.bottom
                )
                // Draw the delete icon
                mTrashIcon.setBounds(
                        itemView.left + iconMargin,
                        iconTop,
                        itemView.left + iconMargin + mTrashIcon.intrinsicWidth,
                        iconBottom
                )
            }

            // Swiping to the left
            dX < 0 -> {
                // Draw background
                mBackground.setBounds(
                        itemView.right + dX.toInt() - backgroundCornerOffset,
                        itemView.top,
                        itemView.right,
                        itemView.bottom
                )
                // Draw the delete icon
                mTrashIcon.setBounds(
                        itemView.right - iconMargin - mTrashIcon.intrinsicWidth,
                        iconTop,
                        itemView.right - iconMargin,
                        iconBottom
                )
            }

            // view is unSwiped
            else -> {
                cornerRadius = 0
                mBackground.setBounds(0, 0, 0, 0)
                mTrashIcon.setBounds(0, 0, 0, 0)
            }
        }

        mBackground.draw(c)
        mTrashIcon.draw(c)

        if (viewHolder.itemView is CardView) {
            val cardView = viewHolder.itemView as CardView
            cardView.radius = cornerRadius.toFloat()
        }
    }

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
        mDeleteListener.onDelete(viewHolder)
    }

    init {
        mBackground = ColorDrawable(Color.RED)
        mTrashIcon = context.resources.getDrawable(R.drawable.ic_baseline_delete_sweep_24, null)
    }
}